test/test_train_mp_mnist.py

```diff
@@ -0,0 +1,160 @@
+        data=(torch.zeros(FLAGS.batch_size, 1, 28,
+                          28), torch.zeros(FLAGS.batch_size,
+                                           dtype=torch.int64)),
+        sample_count=10000 // FLAGS.batch_size // xm.xrt_world_size())
+  else:
+    train_dataset = datasets.MNIST(
+        FLAGS.datadir,
+        train=True,
+        download=True,
+        transform=transforms.Compose(
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    test_dataset = datasets.MNIST(
+        FLAGS.datadir,
+        train=False,
+        transform=transforms.Compose(
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    train_sampler = None
+    if xm.xrt_world_size() > 1:
+      train_sampler = torch.utils.data.distributed.DistributedSampler(
+          train_dataset,
+          num_replicas=xm.xrt_world_size(),
+          rank=xm.get_ordinal(),
+          shuffle=True)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        batch_size=FLAGS.batch_size,
+        sampler=train_sampler,
+        shuffle=False if train_sampler else True,
+        num_workers=FLAGS.num_workers)
+    test_loader = torch.utils.data.DataLoader(
+        test_dataset,
+        batch_size=FLAGS.batch_size,
+        shuffle=False,
+        num_workers=FLAGS.num_workers)
+
+  # Scale learning rate to num cores
+  lr = FLAGS.lr * xm.xrt_world_size()
+
+  device = xm.xla_device()
+  model = MNIST().to(device)
+  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=FLAGS.momentum)
+  loss_fn = nn.NLLLoss()
+
+  def train_loop_fn(loader):
+    tracker = xm.RateTracker()
+
+    model.train()
+    for x, (data, target) in loader:
+      optimizer.zero_grad()
+      output = model(data)
+      loss = loss_fn(output, target)
+      loss.backward()
+      xm.optimizer_step(optimizer)
+      tracker.add(FLAGS.batch_size)
+      if x % FLAGS.log_steps == 0:
+        test_utils.print_training_update(device, x, loss.item(), tracker.rate(),
+                                         tracker.global_rate())
+
+  def test_loop_fn(loader):
+    total_samples = 0
+    correct = 0
+    model.eval()
+    for x, (data, target) in loader:
+      output = model(data)
+      pred = output.max(1, keepdim=True)[1]
+      correct += pred.eq(target.view_as(pred)).sum().item()
+      total_samples += data.size()[0]
+
+    accuracy = 100.0 * correct / total_samples
+    test_utils.print_test_update(device, accuracy)
+    return accuracy
+
+  accuracy = 0.0
+  for epoch in range(1, FLAGS.num_epochs + 1):
+    para_loader = dp.ParallelLoader(train_loader, [device])
+    train_loop_fn(para_loader.per_device_loader(device))
+
+    para_loader = dp.ParallelLoader(test_loader, [device])
+    accuracy = test_loop_fn(para_loader.per_device_loader(device))
+    if FLAGS.metrics_debug:
+      print(torch_xla._XLAC._xla_metrics_report())
+
+  return accuracy
+
+
+def _mp_fn(index, flags):
```

Comment:
can these spawned fn's have a return value? We can probably work around it either way, e.g. by logging and regexing later like we now in nightly testing or asserting in the fn like you do here

Reply:
The mp_fn are really the real main function of the process.
They can exit with different exit status, but we currently do nothing with it.
Good point though. This is needed in order to properly exit with error status.

Reply:
Actually, `torch.multiprocessing.spawn()` in wait-mode, already raises an exception if one of the child exists with a code different from 0.